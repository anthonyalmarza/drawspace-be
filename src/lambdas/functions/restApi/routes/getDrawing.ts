import { API, HandlerFunction } from 'lambda-api'
import validateRequest from 'lambdas/functions/restApi/middleware/validateRequest'
import getDrawing from 'lambdas/util/accessPatterns/getDrawing'

const validteUrlParams = validateRequest(
    {
        drawingId: { presence: true, type: 'string' },
    },
    'params'
)

const handler: HandlerFunction = async (req, res) => {
    const { drawingId } = req.params
    const result = await getDrawing(drawingId)
    res.send(result)
}

export default (api: API): void => {
    api.get('/drawings/:drawingId', validteUrlParams, handler)
}
