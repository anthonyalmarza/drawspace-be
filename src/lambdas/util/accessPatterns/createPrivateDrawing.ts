import drawingEntity from 'lambdas/util/entities/drawing'
import { Drawing, UpdateReturnValues } from 'types'
import ksuid from 'ksuid'

export default async ({
    id = ksuid.randomSync().string,
    title,
    thumbnailUrl,
    drawStepsUrl,
    startTime,
    endTime,
    width,
    height,
    resolution,
    user,
}: {
    id?: string
    title: string
    thumbnailUrl: string
    drawStepsUrl: string
    startTime: string
    endTime: string
    width: number
    height: number
    resolution: number
    user: string
}): Promise<Drawing> => {
    const { Attributes: drawing } = await drawingEntity.update(
        {
            id,
            title,
            thumbnailUrl,
            drawStepsUrl,
            startTime,
            endTime,
            width,
            height,
            resolution,
            user,
        },
        {
            returnValues: UpdateReturnValues.ALL_NEW,
            conditions: { attr: 'pk', exists: false },
        }
    )
    return drawing
}
