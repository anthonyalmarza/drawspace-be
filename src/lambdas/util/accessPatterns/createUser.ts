import userEntity from 'lambdas/util/entities/user'
import { User, UpdateReturnValues } from 'types'

export default async ({
    userId,
    name,
}: {
    userId: string
    name: string
}): Promise<User> => {
    const { Attributes: user } = await userEntity.update(
        { id: userId, name },
        {
            returnValues: UpdateReturnValues.ALL_NEW,
            conditions: { attr: 'pk', exists: false },
        }
    )
    return user
}
