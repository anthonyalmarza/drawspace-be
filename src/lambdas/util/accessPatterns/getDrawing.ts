import drawingEntity from 'lambdas/util/entities/drawing'
import { Drawing } from 'types'

export default async (drawingId: string): Promise<Drawing> => {
    const { Item: drawing } = await drawingEntity.get({ id: drawingId })
    return drawing
}
