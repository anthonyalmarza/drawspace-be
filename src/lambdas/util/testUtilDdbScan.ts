import DynamoDB from 'aws-sdk/clients/dynamodb'

const { apiTableName } = require('lambdas/util/envVarsRestApi')
/* eslint-enable */
const DocumentClient = new DynamoDB.DocumentClient({
    endpoint: 'http://localhost:9001',
    accessKeyId: 'dynamo',
    secretAccessKey: 'devDummyKey',
    region: 'local',
})

export default () => DocumentClient.scan({ TableName: apiTableName }).promise()
