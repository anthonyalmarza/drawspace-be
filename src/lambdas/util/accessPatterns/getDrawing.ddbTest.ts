import { advanceTo } from 'jest-date-mock'
import testUtilDdbScan from 'lambdas/util/testUtilDdbScan'
import createPrivateDrawing from 'lambdas/util/accessPatterns/createPrivateDrawing'
import createUser from 'lambdas/util/accessPatterns/createUser'
import getDrawing from 'lambdas/util/accessPatterns/getDrawing'

advanceTo(new Date(2020, 5, 29, 0, 0, 0))

const mockUserId = '1234123412341234'
const mockDrawingId = '4321432143214321'

test('getDrawing', async () => {
    const drawingMockCommon = {
        title: 'mockTitle',
        thumbnailUrl: 'mockThumbnailUrl',
        drawStepsUrl: 'mockDrawStepsUrl',
        startTime: 'mockStartTime',
        endTime: 'mockEndTime',
        width: 1000,
        height: 1000,
        resolution: 2,
        user: mockUserId,
    }
    await createUser({ userId: mockUserId, name: 'mock user name' })
    await createPrivateDrawing({
        id: mockDrawingId,
        ...drawingMockCommon,
    })

    const res = await getDrawing(mockDrawingId)
    const scan = await testUtilDdbScan()
    expect(res).toMatchSnapshot()
    expect(scan).toMatchSnapshot()
})
