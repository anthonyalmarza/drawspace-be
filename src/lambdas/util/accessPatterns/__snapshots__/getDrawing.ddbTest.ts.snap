// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDrawing 1`] = `
Object {
  "created": "2020-06-29T00:00:00.000Z",
  "drawStepsUrl": "mockDrawStepsUrl",
  "endTime": "mockEndTime",
  "entity": "drawing",
  "height": 1000,
  "id": "4321432143214321",
  "modified": "2020-06-29T00:00:00.000Z",
  "resolution": 2,
  "startTime": "mockStartTime",
  "thumbnailUrl": "mockThumbnailUrl",
  "title": "mockTitle",
  "user": "1234123412341234",
  "width": 1000,
}
`;

exports[`getDrawing 2`] = `
Object {
  "Count": 2,
  "Items": Array [
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#private#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "pk": "drawing#4321432143214321",
      "r": 2,
      "sk": "drawing#4321432143214321",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "user",
      "_md": "2020-06-29T00:00:00.000Z",
      "dc": 0,
      "n": "mock user name",
      "pdc": 0,
      "pk": "user#1234123412341234",
      "sk": "user#1234123412341234",
    },
  ],
  "ScannedCount": 2,
}
`;
